### DISCUSSION TITLE
How to create an empty table based on entity at runtime?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2927] [Q&A]
I need to create some tables at runtime. All tables are under `*` schema (I use PostgreSQL).

Currently, I use the following code to create all the required tables.

```ts
await this.orm.getSchemaGenerator().createSchema({schema: `m${machine}`})
```

However, I need to create different tables in different Nest.js modules in the same schema. The code above does not work, because it always tries to create all the tables defined under the particular schema, however, some of them might already exist and then it will fail to create the missing tables.

I am okay with creating the tables individually, however, I have no idea how to create the tables.

I am looking for a command like this (of course, the following code does not work):

```ts
await this.entityRepository.create(Entity).withSchema({schema: `m${machine}`})

// OR
await this.orm.em.create(Entity).withSchema({schema: `m${machine}`})
```

Of course, I also need to create the schema if it doesn’t exist yet. :wink:

**Update**

I think I am looking for `await this.orm.em.create(Entity, {}, {schema: `m${machine}`})`, however, it does not do anything. Of course, there are some required properties in the entity, but I don’t want to provide any data.

**Comment by B4nan** [comment]
`em.create` is sync method that creates entity instance, it wont do any queries.

**Reply by tukusejssirs** [reply]
Yeah, I left the `await` there by mistake, but it still does not really work. It does not create the table, moreover, `{}` is populated with some properties (columns) which have either the default value set to empty string or are nullable. No error is output.

**Reply by tukusejssirs** [reply]
Basically, all I need to do is create the following commands using MikroORM. I could run these commands using `await this.orm.em.getConnection().execute()`, but I presume there is a better way to do it.

```sql
create schema if not exists m$machine;
create table if not exists m$machine.$table (...);
```

**Reply by tukusejssirs** [reply]
@B4nan, is there a way around inserting a row when I create a table using `await this.orm.em.create(Entity, {}, {schema: m${machine}})`? `{}` is populated with some of the properties (in my test case, it inserts one with an empty string default and another one with `null` default; however, the table is not created, probably because of the missing required properties, but no error is returned).

**Reply by B4nan** [reply]
Once again, `em.create()` is a sync method and serves only for creating entity instances. It will never "create a table", nor it will issue any query - it will only create the entity instance.

You need to use schema generator to work with schema. Sounds like you should be using `updateSchema()` instead of `createSchema()`, as that will respect what tables are already present and sync their state. There is no way to "partially create/update schema", which sounds like what you want. I really don't see any reason to support such thing, as it would literally mean "produce broken schema".

**Reply by tukusejssirs** [reply]
Well, I consider `updateSchema()` a workaround which is not fully suitable for my use case.

I need to create individual tables, as I need to create different tables for different machine vendors (I do a CNC machine monitoring app), e.g. `mx.production_vendor_one` and `mx.production_vendor_two`, as I can get different data from different vendors.

I understand why you see this as a request to produce a broken schema, however, I still wish to be able to create a specific table, e.g., using `createTable()` method in a particular schema.

What could persuade you to add such method to MikroORM?

**Reply by B4nan** [reply]
If you want to create just some tables, dont put others into the ORM config. This way it will also force you to "not produce invalid schema" as it wont let you init the ORM with entities that have incomplete relations - for example if you have book entity with author relation, it wont let you discover only book, as without author there is no way to define the FK.

**Reply by tukusejssirs** [reply]
I don’t think I can leave some entities out. In a single shop/plant, there might be machines of different vendors, however, my app monitors them all. While machine 1 could have `m1.production_vendor_1`, machine 2 could have `m2.machine_vendor_2`. Both tables should be accessible within a single module (`ProductionModule` in Nest.js).

**Reply by tukusejssirs** [reply]
Is there any other way (besides using `updateSchema()`) to create specific tables, while reading the specific entities definition (in order to refrain from hard-coding the table structure in an SQL command)?

**Comment by vis-litash** [comment]
Hi there! 
Facing the same problem with my work project: we have some kind of schema-per-tenant multitenancy in our API service based on PostgreSQL DBMS. The tenants can be different and have their own type, when the tenant is initializing we need to create new schema and only some of tables depending on tenant type. Also there is some "common" tables in default `public` schema.

So to implement such specific logic of dynamic creation of schemas and tables i searched some "low level" public methods in ORM classes, which could be used to control what entities should be passed in schema generator manually or at least generate SQL statements from entity metadata. I've been trying to implement something like that:
```TypeScript
const entityMeta = entityManager.getMetadata().get(MyEntity.name)

const entityDDL = orm.schema /* someMethod */.getQuery(entityMeta)

const newSchemaEntityManager = entityManager.fork({schema: 'new_schema_name'})

await newSchemaEntityManager.execute(entityDDL)

```
But no luck)

In my opinion such "gap" may be very useful in non-standard situations, it may help to avoid duplication of code or using raw SQL and other tricks. Thus i can offer the author of library to consider such functionality not as full-fledged feature, but something like ability to customize the library behavior.

Back to my use-case, for now i ended up with creating of new instance of ORM with modified config to generate schema for another tenant:
```TypeScript
// cfg is Options object imported from mikro-orm.config.ts
async createTenantSchema(schemaName: string, entities: Array<EntityClass<any>>) {
    // explicitly pass new schema name
    cfg.schema = schemaName
    // redefine entities that shoud be generated in new schema
    cfg.entities = entities
    // use separate ORM instance only to generate new schema
    const orm = await MikroORM.init(cfg)
    await orm.schema.createSchema()
    await orm.close()
}
```

Hope this comment would help someone, cheers
